### DISCUSSION TITLE
How to get sender's IP address and port number after io_uring_wait_cqe()?

### DISCUSSION THREAD

**Discussion by cnbatch** [axboe/liburing#581] [Q&A]
When using `recvfrom` or `recvmsg`, the sender's IP address and port number can be retrieved from a `struct sockaddr_in *` or `struct sockaddr_in6 *`.

Now I'm reading the source file `test/send_recvmsg.c` , but I have not find a way to get that address after `io_uring_wait_cqe(ring, &cqe)` in `static int do_recvmsg(struct io_uring *ring, char buf[MAX_MSG + 1], struct recv_data *rd)`.

After calling `io_uring_wait_cqe(),` `cqe->user_data` is still `NULL`.

So how can I get the sender's IP address and port numbers or `struct sockaddr_in *` (or `struct sockaddr_in6 *`) after `io_uring_wait_cqe()`? 

Thanks in advance.

**Comment by ammarfaizi2** [comment]
I personally love the simplicity of the `recvfrom` / `sendto` API to handle
this case.

But io_uring doesn't have them. I wonder why do we support `recv` / `send`,
but not `recvfrom` / `sendto`. Life would be much better if we support the
latter because it's the superset of the former. And we don't actually have
`recv` / `send` syscall in the `net/` anyway.

Yeah, I know, we already have `recvmsg` and `sendmsg`, but the data structure
is overkill for a simple thing like this.

So, you can read `man 2 recvmsg` for the detail. The function prototype of
`recvmsg` is:
```c
ssize_t recvmsg(int sockfd, struct msghdr *msg, int flags);
```

That's what we have in `io_uring_prep_recvmsg()` as well, with the extra first
argument being the SQE to be submitted.

And then we have these two structs:
```c
struct iovec {                    /* Scatter/gather array items */
    void  *iov_base;              /* Starting address */
    size_t iov_len;               /* Number of bytes to transfer */
};

struct msghdr {
    void         *msg_name;       /* optional address */
    socklen_t     msg_namelen;    /* size of address */
    struct iovec *msg_iov;        /* scatter/gather array */
    size_t        msg_iovlen;     /* # elements in msg_iov */
    void         *msg_control;    /* ancillary data, see below */
    size_t        msg_controllen; /* ancillary data buffer len */
    int           msg_flags;      /* flags on received message */
};
```
(Note, the `recvmsg`'s second argument expects the type to be a pointer to
`struct msghdr`).


## TL;DR

By using `recvfrom`, you must be usually doing this:
```c
int fd = ...;
int flags = ...;
struct sockaddr_in addr;
socklen_t len = sizeof(addr);
char buf[N];
ssize_t ret;

ret = recvfrom(fd, buf, sizeof(buf), flags, (struct sockaddr *)&addr, &len);
```

the above call is equivalent to:
```c
int fd = ...;
int flags = ...;
struct sockaddr_in addr;
socklen_t len = sizeof(addr);
char buf[N];
struct msghdr msg;
struct iovec iov;
ssize_t ret;

iov.iov_base = buf;
iov.iov_len = sizeof(buf);
msg.msg_name = &addr;
msg.msg_namelen = len;
msg.msg_iov = &iov;
msg.msg_iovlen = 1;
msg.msg_control = NULL;
msg.msg_controllen = 0;
msg.msg_flags = 0;
ret = recvmsg(fd, &msg, flags);
```
